lib/Differentiator/DiffPlanner.cpp

```diff
@@ -21,38 +21,123 @@ namespace clad {
     struct Finder :
       RecursiveASTVisitor<Finder> {
         DeclRefExpr* m_FnDRE = nullptr;
+        Expr* m_Root;
         Expr** m_RelevantAncestor = nullptr;
-        Finder(Expr** relevantAncestor) : m_RelevantAncestor(relevantAncestor) {}
+        Sema& m_SemaRef;
+        Finder(Sema& SemaRef, Expr* root, Expr** relevantAncestor = nullptr)
+            : m_SemaRef(SemaRef), m_Root(root),
+              m_RelevantAncestor(relevantAncestor) {
+          TraverseStmt(m_Root);
+        }
+
+        // Required for visiting lambda declarations.
+        bool shouldVisitImplicitCode() const { return true; }
+
+        // just required for computing relevantAncestor.
         bool VisitExpr(Expr* E) {
           if (m_RelevantAncestor) {
             switch (E->getStmtClass()) {
               case Stmt::StmtClass::ImplicitCastExprClass: LLVM_FALLTHROUGH;
               case Stmt::StmtClass::UnaryOperatorClass:
                 *m_RelevantAncestor = E;
-                break; 
+                break;
             }
           }
-          if (auto DRE = dyn_cast<DeclRefExpr>(E)) {
-            if (auto VD = dyn_cast<VarDecl>(DRE->getDecl()))
+          return true;
+        }
+
+        bool VisitDeclRefExpr(DeclRefExpr* DRE) {
+          if (auto VD = dyn_cast<VarDecl>(DRE->getDecl())) {
+            auto varType = VD->getType().getTypePtr();
+            // If variable is of class type, set `m_FnDRE` to
+            // `DeclRefExpr` of overloaded call operator method of
+            // the class type.
+            if (varType->isStructureOrClassType()) {
+              auto RD = varType->getAsCXXRecordDecl();
+              TraverseDecl(RD);
+            } else {
               TraverseStmt(VD->getInit());
-            else if (auto FD = dyn_cast<FunctionDecl>(DRE->getDecl()))
-              m_FnDRE = DRE;
-            return false;  
+            }
+          } else if (auto FD = dyn_cast<FunctionDecl>(DRE->getDecl()))
+            m_FnDRE = DRE;
+          return false;
+        }
+
+        bool VisitCXXRecordDecl(CXXRecordDecl* RD) {
+          auto callOperatorDeclName =
+              m_SemaRef.getASTContext().DeclarationNames.getCXXOperatorName(
+                  OverloadedOperatorKind::OO_Call);
+          auto LR = RD->lookup(callOperatorDeclName);
+
+          // Emit error diagnostics
+          if (LR.empty()) {
+            const char diagFmt[] =
+                "No call operator is defined for the class %0";
+            auto diagId =
+                m_SemaRef.Diags.getCustomDiagID(DiagnosticsEngine::Level::Error,
+                                                diagFmt);
+            m_SemaRef.Diag(m_Root->getBeginLoc(), diagId) << RD->getName();
+          } else if (LR.size() > 1) {
+            const char diagFmt[] =
+                "Class %0 defines multiple overloads of call operator. "
+                "Multiple overloads of call operators are not supported.";
```

Comment:
```suggestion
                "'%0' has multiple definitions of operator(). "
                "Multiple definitions of call operators are not supported.";
```

We should probably issue a note for each definition -- see how it is done in clang on ambiguous overload calls.

Reply:
Okay, I will see how it is done in clang on ambiguous overload calls, and try to add it here.

Reply:
Done, now clad issues a note for each candidate function.